pkg/utils/ldap/reconnect.go

```diff
@@ -0,0 +1,341 @@
+	conn, err := c.getConnection()
+	if err != nil {
+		return err
+	}
+
+	for try := 0; try <= c.retries; try++ {
+		err = conn.Del(d)
+		if !ldap.IsErrorWithCode(err, ldap.ErrorNetwork) {
+			// non network error, return it to the client
+			return err
+		}
+
+		c.logger.Debug().Msgf("Network Error. attempt %d", try)
+		conn, err = c.reconnect(conn)
+		if err != nil {
+			return err
+		}
+		c.logger.Debug().Msg("retrying LDAP Del")
+	}
+	// if we get here we reached the maximum retries. So return an error
+	return ldap.NewError(ldap.ErrorNetwork, errMaxRetries)
+}
+
+// Modify implements the ldap.Client interface
+func (c *ConnWithReconnect) Modify(m *ldap.ModifyRequest) error {
+	conn, err := c.getConnection()
+	if err != nil {
+		return err
+	}
+
+	for try := 0; try <= c.retries; try++ {
+		err = conn.Modify(m)
+		if !ldap.IsErrorWithCode(err, ldap.ErrorNetwork) {
+			// non network error, return it to the client
+			return err
+		}
+
+		c.logger.Debug().Msgf("Network Error. attempt %d", try)
+		conn, err = c.reconnect(conn)
+		if err != nil {
+			return err
+		}
+		c.logger.Debug().Msg("retrying LDAP Modify")
+	}
+	// if we get here we reached the maximum retries. So return an error
+	return ldap.NewError(ldap.ErrorNetwork, errMaxRetries)
+}
+
+// ModifyDN implements the ldap.Client interface
+func (c *ConnWithReconnect) ModifyDN(m *ldap.ModifyDNRequest) error {
+	conn, err := c.getConnection()
+	if err != nil {
+		return err
+	}
+
+	for try := 0; try <= c.retries; try++ {
+		err = conn.ModifyDN(m)
+		if !ldap.IsErrorWithCode(err, ldap.ErrorNetwork) {
+			// non network error, return it to the client
+			return err
+		}
+
+		c.logger.Debug().Msgf("Network Error. attempt %d", try)
+		conn, err = c.reconnect(conn)
+		if err != nil {
+			return err
+		}
+		c.logger.Debug().Msg("retrying LDAP ModifyDN")
+	}
+	// if we get here we reached the maximum retries. So return an error
+	return ldap.NewError(ldap.ErrorNetwork, errMaxRetries)
+}
+
+func (c *ConnWithReconnect) getConnection() (*ldap.Conn, error) {
+	conn := <-c.conn
+	if conn.Conn != nil && !ldap.IsErrorWithCode(conn.Error, ldap.ErrorNetwork) {
+		c.logger.Debug().Msg("using existing Connection")
+		return conn.Conn, conn.Error
+	}
+	return c.reconnect(conn.Conn)
+}
+
+func (c *ConnWithReconnect) ldapAutoConnect(config Config) {
+	l, err := c.ldapConnect(config)
+	if err != nil {
+		c.logger.Debug().Err(err).Msg("autoconnect could not get ldap Connection")
+	}
+
+	for {
+		select {
+		case resConn := <-c.reset:
+			// Only close the connection and reconnect if the current
+			// connection, matches the one we got via the reset channel.
+			// If they differ we already reconnected
+			if l != nil && l == resConn {
+				c.logger.Debug().Msgf("closing connection %v", &l)
+				l.Close()
+			}
+			if l == resConn || l == nil {
+				c.logger.Debug().Msg("reconnecting to LDAP")
+				l, err = c.ldapConnect(config)
+			} else {
+				c.logger.Debug().Msg("already reconnected")
+			}
```

Comment:
Uh, this `if` collection is making my head burn :smile: How about refactoring it into a switch?
```golang
switch {
case l == nil:
    l, err = c.ldapConnect(config)
case l != resConn:
    continue
default:
   l.Close()
   l, err = c.ldapConnect(config)
}
```

Reply:
done